Update miscale.py

xMrVizzy opened the pull request:
# Description

Added multiple user profiles and body metrics! (Thank you, https://github.com/lolouk44/xiaomi_mi_scale for Body Metrics!)

New args for metrics and profiles:
<img width="213" alt="CleanShot 2020-01-03 at 00 29 32@2x" src="https://user-images.githubusercontent.com/7906418/71687619-27a02980-2dc0-11ea-8c8e-9ec830d71f19.png">

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

bbbenji commented:
Hey that looks good! I will test it soon. Can you also please update the example config https://github.com/zewelor/bt-mqtt-gateway/blob/master/config.yaml.example

bbbenji commented:
I am thinking it might be better to have a separate worker for this. That way there is one that returns the raw data, and this one, which has some logic. This is a gateway and technically should only relay data, not transform it.

See PR's https://github.com/zewelor/bt-mqtt-gateway/pull/66 & https://github.com/zewelor/bt-mqtt-gateway/pull/67 for a similar situation.

Additionally, this worker should also return the user name and the raw impedance value as well.

Other than that, it's working well.

xMrVizzy commented:
Okay, I will make it now!

xMrVizzy commented:
I made it separate but I need to parse 3 value from miscale worker MQTT topics: weight, impedance, time. Do you know how I can get this values in Python?

xMrVizzy commented:
@bbbenji

zewelor commented:
I would create another worker, or leave it here, but make user config optional. Also it would be much better to support users array, with any numer of users. Adding this feature will make more question about to support single user / three users / more users. It should be somehow written in most possible generic way. Like each user with some weight range + support for unknown weight ?

In general I would prefer to offload such feature for some other software, like HA ( probably appdaemon app ? ).

When another scale will be added here, then it will make more complications how to support it also, with that BMI calculations etc. Another good solution would be to register such calculations as daemon ( look at mysensors ), and make it configurable, to listen for weight topic.

Also please change payload building with strings, You can build dict, it will be serialized to json, when sending message ( example here: https://github.com/zewelor/bt-mqtt-gateway/blob/4da197fa2ea5ed1272b3786e59091b7eacaf39e8/workers/thermostat.py#L80 ).

xMrVizzy commented:
Okay, I think multiple user possible and also make it optional! Soon I will push update!

xMrVizzy commented:
I added users array for multiple users, and made a logic that if users array not defined it will work as default! And also changed payload to json. But I have a question how I can check if key in args exist? Now I made it using try: except:, is there a better way?

zewelor commented on workers/miscale.py:
```diff
@@ -16,14 +19,58 @@
+    
+    def _setup(self):
+        try:
+            if self.users:
```
This can be removed.

zewelor commented on workers/miscale.py:
```diff
@@ -16,14 +19,58 @@
+        if self.is_users:
+            for key, item in self.users.items():
+                greater = float(item['weight_template'].split(' to ')[0])
+                less = float(item['weight_template'].split(' to ')[1])
```
Here you should be able to check using https://www.w3schools.com/python/ref_func_hasattr.asp.

xMrVizzy replied:
Okay

xMrVizzy replied:
It should be hasattr(self, 'users') right?

xMrVizzy replied:
Or use MiscaleWorker as object?

zewelor replied:
Not sure probably self :D I'm not experienced Python programmer, only learning.

zewelor commented:
Looks much cleaner and universal ! Can you allow original repo maintainer access to PR, I would make small cleanups, and it should be ready to merge. ( https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork )

What do you think about changing weight_template to something like:

```
weight_template:
  min: 70
  max: 90
```

I believe it will simplify parsing it ( as now its dependant on format, case senstive and numer of spacer, if someone uses too much etc ). Also it should be easier to make some config format / values validation ( like HA does ). So user has instant feedback, that format is wrong / some keys missing etc. As this config gets more and more comlicated, I think it would be good to move to such solution in future.

xMrVizzy commented:
And I should make a few more changes!

zewelor commented on workers/miscale.py:
```diff
@@ -41,12 +34,9 @@ def status_update(self):
-                less = float(item['weight_template'].split(' to ')[1])
-                
-                if results.weight > greater and results.weight < less:
+                if results.weight > item['weight_template']['min'] and results.weight < item['weight_template']['max']:
```
Maybe ">=" and "<=". I think more people will expect to include range ends, as least suprise

zewelor commented on workers/miscale.py:
```diff
@@ -1,3 +1,6 @@
+from math import floor
+import json
```
I think this can be removed, mqtt lib should handle serialization on its own.

xMrVizzy replied:
You mean json.dump()?

xMrVizzy replied:
If it, MQTT serialize payload only when using MqttMessageConfig, in MqttMessage there is no serialization and it send error that payload should be string, not object!

xMrVizzy replied:
Screen of MqttMessageConfig:
<img width="597" alt="CleanShot 2020-01-04 at 16 56 21@2x" src="https://user-images.githubusercontent.com/7906418/71765175-259aaf80-2f13-11ea-8750-422fdccd88fb.png">

Screen of MqttMessage:
<img width="489" alt="CleanShot 2020-01-04 at 16 56 43@2x" src="https://user-images.githubusercontent.com/7906418/71765180-32b79e80-2f13-11ea-85e6-2ffcf22da668.png">

zewelor replied:
Oh you are right. I need to write this better, probably MqttMessage should check message type and do json dump on his own, instead of need to do it in each place. I can clear that later.

zewelor commented on workers/miscale.py:
```diff
@@ -16,14 +19,48 @@
+                        metrics['protein'] = float("{:.2f}".format(lib.getProteinPercentage()))
+                    
+                    metrics['timestamp'] = results.midatetime
+                    messages.append(MqttMessage(topic=self.format_topic("users/" + user), payload=json.dumps(metrics)))
```
I think this can be removed, mqtt lib should handle serialization on its own. Just pass metrics here

xMrVizzy commented:
Okay, I tested and all work!

zewelor commented:
Also manually tested all cases, no users, no user withing range, user withing range. All works fine. Merging this. Thanks for great PR !

xMrVizzy commented:
Thank you!
